File: working hours register/Level 2/worker_service.py
import heapq

class WorkerService:
    def __init__(self, repo: WorkerRepository):
        self.repo = repo

    def add_worker(self, worker_id: str, position: str, compensation: int) -> str:
        return self.repo.add_worker(worker_id, position, compensation)

    def register(self, worker_id: str, timestamp: int) -> str:
        worker = self.repo.get_worker(worker_id)
        return worker.register(timestamp) if worker else "invalid_request"

    def get_time(self, worker_id: str) -> str:
        worker = self.repo.get_worker(worker_id)
        return worker.get_total_minutes() if worker else ""

    def top_n_workers(self, n: int, position: str) -> str:
        workers = self.repo.get_workers_by_position(position)
        if not workers:
            return ""

        heap = []

        for worker in workers:
            # Create a max-heap using (-time, worker_id) to get proper order
            heapq.heappush(heap, (-worker.total_minutes, worker.worker_id, worker))

            if len(heap) > n:
                heapq.heappop(heap)

        # Now extract and sort the heap to build the result in proper descending order
        top_n = sorted(heap, key=lambda x: (-x[0], x[1]))  # sort by time desc, id asc

        return ", ".join(f"{worker.worker_id}({-score})" for score, _, worker in top_n)

File: working hours register/Level 2/C#/workerService.cs
// WorkerService.cs
using System.Collections.Generic;
using System.Linq;

public class WorkerService
{
    private readonly WorkerRepository _repo;

    public WorkerService(WorkerRepository repo)
    {
        _repo = repo;
    }

    public string AddWorker(string workerId, string position, int compensation)
    {
        return _repo.AddWorker(workerId, position, compensation);
    }

    public string Register(string workerId, int timestamp)
    {
        var worker = _repo.GetWorker(workerId);
        return worker != null ? worker.Register(timestamp) : "invalid_request";
    }

    public string GetTime(string workerId)
    {
        var worker = _repo.GetWorker(workerId);
        return worker != null ? worker.GetTotalMinutes() : "";
    }

    public string TopNWorkers(int n, string position)
    {
        var workers = _repo.GetWorkersByPosition(position);
        if (!workers.Any()) return "";

        var minHeap = new SortedSet<(int totalMinutes, string workerId, Worker worker)>(
            new WorkerEntryComparer()
        );

        foreach (var worker in workers)
        {
            var entry = (worker.TotalMinutes, worker.WorkerId, worker);
            minHeap.Add(entry);

            if (minHeap.Count > n)
            {
                // Remove the one with the smallest totalMinutes or largest workerId
                minHeap.Remove(minHeap.Max);
            }
        }

        var result = minHeap
            .OrderByDescending(x => x.totalMinutes)
            .ThenBy(x => x.workerId)
            .Select(x => $"{x.workerId}({x.totalMinutes})");

        return string.Join(", ", result);
    }
}
